[Draft] Define stages for changes to GraphQL specification

IvanGoncharov opened the pull request:
This PR is a follow-up WG discussion about defining stages for changes to GraphQL specification. 

I copied stage descriptions from [my slides](https://docs.google.com/presentation/d/1y7OLdMv4sL583vUGePcXMMvmYRGz7JlBowaYDMcHWCU/edit?usp=sharing) as an early draft to keep momentum and continue the conversation.

wincent commented on stages.md:
```diff
@@ -0,0 +1,49 @@
+## Stage 2: Candidate
+**Prerequisite**:
++ Prepared PR for graphql-js
++ Notify all members of GraphQL WG
```
Should probably link to the WG repo here.

wincent commented on stages.md:
```diff
@@ -0,0 +1,49 @@
+## Stage -2: proposed change(optional)
+**Prerequisite**: Described problem/change should be specific to the content of GraphQL Specification and not be an implementation detail.
+
+**Purpose**: Filter out question, issues for other repos and engage community discussion.
```
question → questions

wincent commented on stages.md:
```diff
@@ -0,0 +1,49 @@
+## Stage 0: Proposal
+**Prerequisite**:
++ Initial version of spec changes
++ Filled checklist in PR description
```
What do you have in mind for the checklist?

wincent commented on stages.md:
```diff
@@ -0,0 +1,49 @@
+## Stage 1: Draft
+**Prerequisite**:
++ Finalized wording inside Specification document
++ Proposed spec changes don’t have any blind spots (not described edge-cases, missed changes to related part of spec, etc.)
```
not described → undescribed

wincent commented on stages.md:
```diff
@@ -0,0 +1,49 @@
++ Finalized wording inside Specification document
++ Proposed spec changes don’t have any blind spots (not described edge-cases, missed changes to related part of spec, etc.)
+
+**Steps**: start working on “graphql-js” PR
```
As this is the first mention of graphql-js in the document, should probably link to the repo.

IvanGoncharov commented:
@wincent Thank you for proofreading 👍

> I think the one thing that is missing here is a framing statement at the top to provide a little more context

I've added an one based on the text from your commit. What do you think?

> What do you have in mind for the checklist?

I think [Lee's list](https://youtu.be/mePT9MNTM98?t=20m32s) from GraphQL Europe talk can serve as a basis. So in theory, it will stimulate PR author to answer basic questions before making PR and thus save reviewers time and improve quality of PRs.

But before adding something concrete I want to make small research and see how it's done in other WG, committee, etc. I added `TBD` not to block discussion of stages since I think we can define checklist afterward.

wincent commented on stages.md:
```diff
@@ -0,0 +1,70 @@
++ Encourage community discussion around the proposed changes.
+
+Any changes to a specification that affect behavior of **GraphQL server or
+client implementation** should go through bellow stages no matter how small
```
bellow → below

wincent commented on stages.md:
```diff
@@ -0,0 +1,70 @@
+it is. All other changes should be labeled as “editorial” and could be merged
+right away.
+
+## Stage -2: proposed change(optional)
```
I think we want whitespace before "(optional)" here, right? Same for the next stage heading as well.

wincent commented:
Idea: perhaps this file should be named `CONTRIBUTING.md` (matching a pattern used by a lot of projects, and [some of GitHub's tooling makes use of it automatically](https://github.com/blog/1184-contributing-guidelines)).

IvanGoncharov commented:
@wincent Great idea 💡 
✅ Done

wincent commented:
So, with this we have a draft definition of "stages for changes to GraphQL specification".

What we don't have is a definition of "stages for changes to the definition of stages for changes to the GraphQL specification".

My inclination is to merge this as-is (but add a "Draft" label) and iterate from there.

IvanGoncharov commented:
> My inclination is to merge this as-is (but add a "Draft" label) and iterate from there.

@wincent I suggest we proceed toward `Candidate` stage by:
+ Notify all members of [GraphQL WG](https://github.com/graphql/graphql-wg)
+ Community consent on the proposed change. If it’s hard to achieve, add it to
agenda of the next WG meeting.
+ No changes to the graphql PR for at least last 7 days

What do you think?

wincent commented:
I'd say we can afford to move a little more quickly/scrappily on merging a change like this: it is not a change to the spec itself, it requires no implementation work from library or product developers, it is easily revised once landed, and you've *already* canvassed the idea in the [last (first?) WG meeting](https://github.com/graphql/graphql-wg/blob/master/notes/2017-08-14.md) to broad (informal) approval. I wouldn't want something like this PR to get bogged down too heavily in process, especially when we can make the non-final nature of the document clear by labeling it as a draft (and at that point I think it *would* make sense to discuss it in detail at the next WG meeting).

IvanGoncharov commented:
@wincent Make sense 👍 
Should I add something to the document itself? For example:
`**Draft document**:  discussion [here](https://github.com/facebook/graphql/pull/342).`

wincent commented:
> Should I add something to the document itself?

Yeah, that's almost exactly what I was thinking.

```
> **Document status:** [Draft](https://github.com/facebook/graphql/pull/342)
```

(Blockquote to make it stand out a bit.)

IvanGoncharov commented:
@wincent ✅ Done

wincent commented:
Ok, great. This is a good starting point. We can iterate from here. Thanks @IvanGoncharov.

wincent commented:
Added discussion of this to the agenda for the next meeting: https://github.com/graphql/graphql-wg/pull/22
